### PR TITLE
Implement most duration, date & time related functions/operators

### PR DESCRIPTION
**Pull request by ThomasGravekamp**
- Implement duration datatype as stated in the spec (split duration into yearMonthDuration and dayTimeDuration)
- Add component extraction functions for duration and date & time types
- Let date & time types use dayTimeDuration for keeping track of timezones
- Add contextual date & time functions
- Add compares for date & time types
- Add compares for duration types
- Add arithmetic operators for durations
- Number of small refactors/fixes

### CODE REVIEW

File: src/selectors/functions/builtInFunctions.numeric.js [hunk 1]
```diff
@@ -115,18 +115,18 @@ function fnRound (halfToEven, _dynamicContext, sequence, precision) {
...
-function fnNumber (_dynamicContext, sequence) {
+function fnNumber (dynamicContext, arg = undefined) {
+	const sequence = arg ? arg : dynamicContext.contextItem;
+
 	if (sequence.isEmpty()) {
 		return Sequence.singleton(createAtomicValue(NaN, 'xs:double'));
 	}
+
 	try {
 		return Sequence.singleton(castToType(sequence.first(), 'xs:double'));
```
Comment (DrRataplan) [inline src/selectors/functions/builtInFunctions.numeric.js]: I'd say it would be better to directly use tryCastToType, and check for errors here.

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
+class DayTimeDuration {
+	constructor (seconds) {
+		// if (seconds > Number.MAX_SAFE_INTEGER || Math.abs(seconds) === Infinity) {
```
Comment (DrRataplan) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: commented code

File: src/selectors/operators/numeric/BinaryNumericOperator.js [hunk 3]
```diff
@@ -58,32 +61,77 @@ class BinaryNumericOperator extends Selector {
...
+		if (isInstanceOfType(firstValue, 'xs:yearMonthDuration') && isInstanceOfType(secondValue, 'xs:yearMonthDuration')) {
+			if (this._kind === '+') {
+				typedResult = createAtomicValue(YearMonthDuration.add(firstValue.value, secondValue.value), 'xs:yearMonthDuration');
+			}
+			if (this._kind === '-') {
+				typedResult = createAtomicValue(YearMonthDuration.subtract(firstValue.value, secondValue.value), 'xs:yearMonthDuration');
+			}
+			if (this._kind === 'div') {
+				typedResult = createAtomicValue(YearMonthDuration.divideByYearMonthDuration(firstValue.value, secondValue.value), 'xs:decimal');
+			}
```
Comment (DrRataplan) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: What if the operator was '*'?

File: src/selectors/operators/numeric/BinaryNumericOperator.js [hunk 4]
```diff
@@ -58,32 +61,77 @@ class BinaryNumericOperator extends Selector {
...
+		if (!typedResult) {
+			if (isInstanceOfType(firstValue, 'xs:untypedAtomic')) {
+				firstValue = castToType(firstValue, 'xs:double');
+			}
+
+			if (isInstanceOfType(secondValue, 'xs:untypedAtomic')) {
+				secondValue = castToType(secondValue, 'xs:double');
+			}
+
+			if (!isInstanceOfType(firstValue, 'xs:numeric') || !isInstanceOfType(secondValue, 'xs:numeric')) {
```
Comment (DrRataplan) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: Numerics wil occur way more, could you pull this to the top?
This file will have to be rewritten in the rebase (assuming #6 will land earlier)

File: test/specs/parsing/functions/builtinFunctions.context.tests.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+let documentNode;
+beforeEach(() => {
+	documentNode = slimdom.createDocument();
+});
+
+describe('Context related functions', () => {
+	describe('fn:current-dateTime', () => {
+		it('returns the current dateTime',
+			() => chai.assert.isOk(evaluateXPathToString('current-dateTime()', documentNode)));
```
Comment (DrRataplan) [inline test/specs/parsing/functions/builtinFunctions.context.tests.js]: Could you check if the current-dateTime returns the same for an execution, but something else after a while?

File: src/selectors/DynamicContext.js [hunk 6]
```diff
@@ -43,6 +46,19 @@ class DynamicContext {
...
+		 * @type {DateTime}
+		 * @const
+		 */
+		this.currentDateTime = DateTime.fromString(date.toISOString());
+
+		/**
+		 * @type {DayTimeDuration}
+		 * @const
+		 */
+		this.implicitTimezone = DayTimeDuration.fromJavascriptDateTimezone(date);
```
Comment (DrRataplan) [inline src/selectors/DynamicContext.js]: I'd say it is better to hard-code it to GMT

File: src/selectors/functions/builtInFunctions.context.js [hunk 7]
```diff
@@ -17,6 +17,22 @@ function fnPosition (dynamicContext) {
 	return Sequence.singleton(createAtomicValue(dynamicContext.contextItemIndex + 1, 'xs:integer'));
 }
 
+function fnCurrentDateTime (dynamicContext) {
+	return Sequence.singleton(createAtomicValue(dynamicContext.currentDateTime, 'xs:dateTimeStamp'));
+}
+
+function fnCurrentDate (dynamicContext) {
+	return Sequence.singleton(createAtomicValue(dynamicContext.currentDateTime.clone().convertToType('xs:date'), 'xs:date'));
```
Comment (DrRataplan) [inline src/selectors/functions/builtInFunctions.context.js]: why clone it?

File: src/selectors/operators/compares/valueCompare.js [hunk 8]
```diff
@@ -1,24 +1,113 @@
...
-    }
+		throw new Error('XPTY0004: Sequences to compare are not singleton');
+	}
+
+	let firstValue = firstSequence.first();
+	let secondValue = secondSequence.first();
+
+	const bothInstanceOfType = type => isInstanceOfType(firstValue, type) && isInstanceOfType(secondValue, type);
+
+	if (bothInstanceOfType('xs:dateTime') ||
```
Comment (DrRataplan) [inline src/selectors/operators/compares/valueCompare.js]: In the future, it is better to structure your conditionals to make the most common case the first. ie: first numerics, then untypedAtomic, etc, etc, etc. This will have to be re-written before rebasing.

File: test/unrunnableTestCases.csv [hunk 9]
```diff
@@ -1352,115 +1300,43 @@ op-concatnni2args-2,Error: FOCA0003: can not cast 303884545991464527 to xs:integ
...
-K-DayTimeDurationEQ-5,AssertionError: Expected XPath not(xs:dayTimeDuration("P3DT08H34M12.143S") ne xs:dayTimeDuration("P3DT08H34M12.143S")) to resolve to true: expected false to be true
-K-DayTimeDurationEQ-6,AssertionError: Expected XPath xs:dayTimeDuration("-PT0S") eq xs:dayTimeDuration("PT0S") to resolve to true: expected false to be true
-K-DayTimeDurationEQ-8,AssertionError: Expected XPath xs:dayTimeDuration("P6DT16H34M59.613999S") eq xs:dayTimeDuration("P6DT16H34M59.613999S") to resolve to true: expected false to be true
-K-YearMonthDurationEQ-1,AssertionError: Expected XPath xs:yearMonthDuration("P1999Y10M") eq xs:yearMonthDuration("P1999Y10M") to resolve to true: expected false to be true
-K-YearMonthDurationEQ-2,AssertionError: Expected XPath xs:yearMonthDuration("P0M") eq xs:yearMonthDuration("P0M") to resolve to true: expected false to be true
-K-YearMonthDurationEQ-5,AssertionError: Expected XPath not(xs:yearMonthDuration("P1999Y10M") ne xs:yearMonthDuration("P1999Y10M")) to resolve to true: expected false to be true
+K-DurationEQ-15,TypeError: Cannot read property 'w' of undefined
```
Comment (DrRataplan) [inline test/unrunnableTestCases.csv]: typeerror, looks like a bug

File: test/unrunnableTestCases.csv [hunk 10]
```diff
@@ -1352,115 +1300,43 @@ op-concatnni2args-2,Error: FOCA0003: can not cast 303884545991464527 to xs:integ
...
 K-HexBinaryEQ-1,AssertionError: Expected XPath xs:hexBinary("FF") eq xs:hexBinary("ff") to resolve to true: expected false to be true
 K-HexBinaryEQ-4,AssertionError: Expected XPath xs:hexBinary(xs:base64Binary(xs:hexBinary("03"))) eq xs:hexBinary("03") to resolve to true: expected false to be true
 fn-intersect-node-args-003,Error: XPST0017: Function root with arity of 1 not registered. No similar functions found.
 fn-intersect-node-args-007,Error: XPTY0004 Unable to convert to type
 fn-intersect-node-args-015,AssertionError: Expected XPath  (/atomic:root/atomic:integer) intersect (/atomic:root/atomic:integer) to resolve to the given XML: expected [ Array(1) ] to deeply equal [ Array(1) ]
 fn-intersect-node-args-016,AssertionError: Expected XPath (/atomic:root/atomic:integer) intersect (/atomic:root/atomic:integer) to resolve to the given XML: expected [ Array(1) ] to deeply equal [ Array(1) ]
+K-DayTimeDurationMultiply-1,AssertionError: Expected XPath xs:dayTimeDuration("P3DT4H3M3.100S") * 3 eq xs:dayTimeDuration("P9DT12H9M9.3S") to resolve to true: expected false to be true
+K-DayTimeDurationMultiply-2,Error: FOCA0005: Cannot divide xs:dayTimeDuration by NaN
```
Comment (DrRataplan) [inline test/unrunnableTestCases.csv]: Shouldn't be NaN?

File: test/unrunnableTestCases.csv [hunk 11]
```diff
@@ -2064,11 +1914,9 @@ cbcl-castable-dateTime-002,Error: FODT0001: Datetime year is out of bounds
...
 cbcl-castable-duration-001,AssertionError: Expected XPath "-P768614336404564651Y" castable as xs:duration to resolve to false: expected true to be false
-cbcl-castable-duration-002,AssertionError: Expected XPath "P768614336404564651Y" castable as xs:duration to resolve to false: expected true to be false
-cbcl-castable-name-002,Error: XPST0017: Function fn:current-time with arity of 0 not registered. No similar functions found.
-cbcl-castable-ncname-002,Error: XPST0017: Function fn:current-time with arity of 0 not registered. No similar functions found.
+cbcl-castable-duration-002,Error: FODT0002: Value overflow while constructing xs:yearMonthDuration
 cbcl-castable-yearMonthDuration-001,AssertionError: Expected XPath "-P768614336404564651Y" castable as xs:yearMonthDuration to resolve to false: expected true to be false
-cbcl-castable-yearMonthDuration-002,AssertionError: Expected XPath "P768614336404564651Y" castable as xs:yearMonthDuration to resolve to false: expected true to be false
+cbcl-castable-yearMonthDuration-002,Error: FODT0002: Value overflow while constructing xs:yearMonthDuration
```
Comment (DrRataplan) [inline test/unrunnableTestCases.csv]: The test case is 
`   <test-case name="cbcl-castable-yearMonthDuration-002">
      <description> test castable to xs:yearMonthDuration with overflow </description>
      <created by="Tim Mills" on="2008-04-25"/>
      <test>"P768614336404564651Y" castable as xs:yearMonthDuration</test>
      <result>
         <assert-false/>
      </result>
   </test-case>
`

Looks like we should return false, not throw an error?

**Comment by DrRataplan** [comment]
Could you also look into the failing tests? They might have to do with inconsistent timezone implementation

File: src/selectors/operators/numeric/BinaryNumericOperator.js [hunk 12]
```diff
@@ -4,21 +4,24 @@ import Sequence from '../../dataTypes/Sequence';
...
-        case '*':
-            return a * b;
-        case 'div':
-            return a / b;
-        case 'idiv':
-            return Math.trunc(a / b);
-        case 'mod':
-            return a % b;
-    }
+	switch (kind) {
```
Comment (bwrrp) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: Is there a particular reason for all binary operators to run through these switches? Can't we distinguish the kind of operator at parse-time and just implement each kind as its own thing?

Comment (bwrrp) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: Alternatively, this seems very similar to calling functions. The approach taken there (data-driven type checks + dispatch) could possibly work here, instead of the massive if / switch blocks.

File: test/unrunnableTestCases.csv [hunk 13]
```diff
@@ -95,12 +95,9 @@ fn-avg-6,Error: XPTY0004: expected argument list of function concat to be [item(
...
 cbcl-avg-006,Error: FORG0006: items passed to fn:avg are not all numeric.
-cbcl-avg-010,AssertionError: Expected executing the XPath "avg((xs:double("NaN"), current-date() - xs:date("1997-01-01") ))" to resolve to one of the expected results, but got AssertionError: expected [Function] to throw error including 'FORG0006' but got 'XPST0017: Function current-date with arity of 0 not registered. No similar functions found.', Error: XPST0017: Function current-date with arity of 0 not registered. No similar functions found..
-cbcl-avg-011,Error: XPST0017: Function day-from-date with arity of 1 not registered. No similar functions found.
-cbcl-avg-012,Error: XPST0017: Function day-from-date with arity of 1 not registered. No similar functions found.
+cbcl-avg-010,AssertionError: Expected executing the XPath "avg((xs:double("NaN"), current-date() - xs:date("1997-01-01") ))" to resolve to one of the expected results, but got AssertionError: expected [Function] to throw error including 'FORG0006' but got 'XPTY0004: the operands of the "-" operator should be of type xs:numeric?.', Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?..
```
Comment (bwrrp) [inline test/unrunnableTestCases.csv]: Shouldn't this work now?

File: test/unrunnableTestCases.csv [hunk 14]
```diff
@@ -244,9 +240,30 @@ fn-countnpi1args-1,Error: FOCA0003: can not cast -999999999999999999 to xs:integ
...
+fn-current-time-5,Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?.
+fn-current-time-6,Error: XPTY0004: the operands of the "+" operator should be of type xs:numeric?.
+fn-current-time-7,Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?.
+fn-current-time-9,Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?.
+fn-dateTime-22,Error: XPST0017: Function fn:adjust-dateTime-to-timezone with arity of 2 not registered. No similar functions found.
+fn-dateTime-23,Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?.
+fn-dateTime-24,Error: XPTY0004: the operands of the "+" operator should be of type xs:numeric?.
+fn-dateTime-25,Error: XPTY0004: the operands of the "+" operator should be of type xs:numeric?.
+fn-dateTime-26,Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?.
+fn-dateTime-27,Error: XPTY0004: the operands of the "-" operator should be of type xs:numeric?.
```
Comment (bwrrp) [inline test/unrunnableTestCases.csv]: Same

File: test/unrunnableTestCases.csv [hunk 15]
```diff
@@ -1304,6 +1251,7 @@ xs-error-047,AssertionError: expected [Function] to throw error including 'FOER0
 xs-error-048,AssertionError: expected [Function] to throw error including 'XPDY0050' but got 'XPST0003: Unable to parse XPath: 1 treat as xs:error. SyntaxError: Expected " ", "!", "!=", "(:", "*", "+", ",", "-", "/", "//", "<", "<<", "<=", "=", "=>", ">", ">=", ">>", "\\n", "\\r", "\\t", "and", "cast", "castable", "div", "eq", "except", "ge", "gt", "idiv", "instance", "intersect", "is", "le", "lt", "mod", "ne", "or", "to", "union", "|", "||", or end of input but "t" found.'
 xs-error-049,Error: XPST0003: Unable to parse XPath: (() treat as xs:error?) => empty(). SyntaxError: Expected " ", "!", "!=", "(:", ")", "*", "+", ",", "-", "/", "//", "<", "<<", "<=", "=", "=>", ">", ">=", ">>", "\n", "\r", "\t", "and", "cast", "castable", "div", "eq", "except", "ge", "gt", "idiv", "instance", "intersect", "is", "le", "lt", "mod", "ne", "or", "to", "union", "|", or "||" but "t" found.
 xs-float-004,AssertionError: expected [Function] to throw an error
+K-DayTimeDurationAdd-3,AssertionError: Expected XPath xs:dayTimeDuration("P3DT4H3M3.100S") + xs:dayTimeDuration("P3DT12H31M56.303S") eq xs:dayTimeDuration("P6DT16H34M59.403S") to resolve to true: expected false to be true
```
Comment (bwrrp) [inline test/unrunnableTestCases.csv]: Precision issue?

File: test/unrunnableTestCases.csv [hunk 16]
```diff
@@ -1352,115 +1300,43 @@ op-concatnni2args-2,Error: FOCA0003: can not cast 303884545991464527 to xs:integ
...
-op-dayTimeDuration-equal2args-12,AssertionError: Expected XPath xs:dayTimeDuration("P15DT11H59M59S") le xs:dayTimeDuration("P0DT0H0M0S") to resolve to false: expected true to be false
-op-dayTimeDuration-equal2args-13,AssertionError: Expected XPath xs:dayTimeDuration("P31DT23H59M59S") le xs:dayTimeDuration("P0DT0H0M0S") to resolve to false: expected true to be false
-op-dayTimeDuration-equal2args-19,AssertionError: Expected XPath xs:dayTimeDuration("P0DT0H0M0S") ge xs:dayTimeDuration("P15DT11H59M59S") to resolve to false: expected true to be false
-op-dayTimeDuration-equal2args-20,AssertionError: Expected XPath xs:dayTimeDuration("P0DT0H0M0S") ge xs:dayTimeDuration("P31DT23H59M59S") to resolve to false: expected true to be false
+op-divide-dayTimeDuration2args-2,AssertionError: xs:dayTimeDuration("P15DT11H59M59S") div xs:double("-1.7976931348623157E308"): expected '-PT7.449541715597465e-303S' to equal 'PT0S'
+op-divide-dayTimeDuration2args-3,AssertionError: xs:dayTimeDuration("P31DT23H59M59S") div xs:double("-1.7976931348623157E308"): expected '-PT1.5379704947317132e-302S' to equal 'PT0S'
```
Comment (bwrrp) [inline test/unrunnableTestCases.csv]: is scientific notation allowed in the duration format?

File: test/unrunnableTestCases.csv [hunk 17]
```diff
@@ -2338,19 +2119,18 @@ ForExpr021,AssertionError: for $i in (1, 2), $j in (3, 4) return ($i, $j): expec
...
 K-ForExprWithout-27,AssertionError: expected [Function] to throw error including 'XPST0081' but got 'XPST0008, The variable PREFIXNOTEXIST:NOTEXIST is not in scope.'
 K-ForExprWithout-30,AssertionError: expected [Function] to throw error including 'XPST0081' but got 'XPST0008, The variable prefix:NOTEXIST is not in scope.'
 K-ForExprWithout-33,AssertionError: expected [Function] to throw error including 'XPST0081' but got 'XPST0008, The variable PREFIXNOTEXIST:NOTEXIST is not in scope.'
 K-ForExprWithout-36,AssertionError: expected [Function] to throw an error
 K-ForExprWithout-37,AssertionError: expected [Function] to throw an error
-K-ForExprWithout-55,Error: XPST0017: Function current-time with arity of 0 not registered. No similar functions found.
-K-ForExprWithout-56,Error: XPST0017: Function timezone-from-time with arity of 1 not registered. No similar functions found.
+K-ForExprWithout-55,AssertionError: Expected XPath count(for $i in (1, 2, current-time()) return ($i, $i)) eq 6 to resolve to true: expected false to be true
```
Comment (bwrrp) [inline test/unrunnableTestCases.csv]: error in for expression? or doesn't current-time() return a singleton sequence?

File: src/selectors/DynamicContext.js (hunk 6 above)
Comment (bwrrp) [inline src/selectors/DynamicContext.js]: Sounds good, maybe make it configurable later?

File: test/unrunnableTestCases.csv (hunk 13 above)
Comment (ThomasGravekamp) [inline test/unrunnableTestCases.csv]: Nope, no arithmetic operators implemented for datetime types.

File: test/unrunnableTestCases.csv (hunk 14 above)
Comment (ThomasGravekamp) [inline test/unrunnableTestCases.csv]: Same

File: test/unrunnableTestCases.csv (hunk 15 above)
Comment (ThomasGravekamp) [inline test/unrunnableTestCases.csv]: Yep, comes down to <some_big_value_in_seconds>.403 % 60 = <number_of_seconds_to_display>.4029999975.....

File: src/selectors/operators/numeric/BinaryNumericOperator.js (hunk 12 above)
Comment (DrRataplan) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: First comment: Nope, XPath is a dynamically typed language;
`let $a := random-number-generator()('permute')(( xs:dateTime#1, xs:string#1))("0") return $a + 1`
:no_good_man: 

Second comment:
rather something like #6, CompareExpression: precomputing the function we will need to execute.

File: test/unrunnableTestCases.csv (hunk 16 above)
Comment (ThomasGravekamp) [inline test/unrunnableTestCases.csv]: Nope, issue with precision. Every integer above Number.MAX_SAFE_INTEGER and below Number.MIN_SAFE_INTEGER will throw FOTD0002, however, a check for (-)0.12345e-300 should be implemented.

**Comment by DrRataplan** [comment]
Could you look into why the CI is f(l)ailing? Could be something with timezones.

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js (hunk 2 above)
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: Fixed

File: test/unrunnableTestCases.csv (hunk 17 above)
Comment (DrRataplan) [inline test/unrunnableTestCases.csv]: looks like it, could be a bug in the ForExpression?

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+		// 	throw new Error('FODT0002: Value overflow while constructing xs:yearMonthDuration');
+		// }
+		// this._seconds = seconds < Number.MIN_SAFE_INTEGER || Object.is(-0, seconds) ? 0 : seconds;
```
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: Overflow check moved to the multiply and division functions.

File: src/selectors/operators/numeric/BinaryNumericOperator.js (hunk 12 above)
Comment (bwrrp) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: A + would still be a + though? Although the type checks are similar for all operators, even if you can't easily separate out the overloads, it seems you could at least separate the operators themselves?

Comment (DrRataplan) [inline src/selectors/operators/numeric/BinaryNumericOperator.js]: Ah, I did not understand you. Yeah, that is better. To prevent making six different constructors, could you make it like CompareExpression? That one builds a function which executes a compare, including the necessary casts.

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+	getMonths () {
+		return 0;
+	}
+
+	getDays () {
+		return (this._seconds / 86400) | 0;
+	}
+
+	getHours () {
+		return (this._seconds % 86400 / 3600) | 0;
```
Comment (DrRataplan) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: Why not use Math.trunc ? I reckon its a bit easier to read then  | 0.

File: test/specs/parsing/functions/builtinFunctions.duration.tests.js [hunk 20]
```diff
@@ -86,14 +86,14 @@ describe('Duration related functions', () => {
 	});
 
 	describe('fn:seconds-from-duration', () => {
-		it('returns 10 for "P10Y10M10DT10H10M10.1S"',
```
Comment (DrRataplan) [inline test/specs/parsing/functions/builtinFunctions.duration.tests.js]: fractional seconds are not important?

File: src/selectors/dataTypes/casting/castToDayTimeDuration.js [hunk 21]
```diff
@@ -1,12 +1,12 @@
...
 
 const createDayTimeDurationValue = value => createAtomicValue(value, 'xs:dayTimeDuration');
 
 /**
  * @param  {./AtomicValueDataType}  value
  * @param  {function(string):boolean}  instanceOf
- * @return {{successful: boolean, value: ../valueTypes/Duration}|{successful: boolean, error: !Error}}
+ * @return {{successful: boolean, value: ../valueTypes/DayTimeDuration}|{successful: boolean, error: !Error}}
  */
 export default function castToDayTimeDuration (value, instanceOf) {
```
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/casting/castToDayTimeDuration.js]: Done

File: src/selectors/dataTypes/valueTypes/QName.js [hunk 22]
```diff
@@ -1,23 +0,0 @@
-class QName {
```
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/QName.js]: Leftover from implementing data types, until now not used anywhere, just some cleanup.

File: src/selectors/DynamicContext.js (hunk 6 above)
Comment (ThomasGravekamp) [inline src/selectors/DynamicContext.js]: Done

File: src/selectors/functions/builtInFunctions.context.js (hunk 7 above)
Comment (ThomasGravekamp) [inline src/selectors/functions/builtInFunctions.context.js]: Fixed

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js (hunk 19 above)
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: Fixed

File: src/selectors/dataTypes/valueTypes/Duration.js [hunk 23]
```diff
@@ -13,190 +16,192 @@ const monthsToMinMaxValues = [
...
 
-function dayTimeDurationToString (duration) {
-	const stringValue = `${duration._days ? `${duration._days}DT` : 'T'}` +
-		`${duration._hours ? `${duration._hours}H` : ''}` +
-		`${duration._minutes ? `${duration._minutes}M` : ''}` +
-		`${duration._seconds || duration._secondFraction ? `${duration._seconds + duration._secondFraction}S` : ''}`;
+function computeMaxDays (duration) {
+	const years = Math.abs(duration.getYears());
+	const months = Math.abs(duration.getMonths());
+	const maxNumberOfLeapYears = Math.floor(years / 4) + (years % 4 > 0 ? 1 : 0);
```
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/Duration.js]: Fixed

File: src/selectors/functions/builtInFunctions.context.js (hunk 7 above)
Comment (ThomasGravekamp) [inline src/selectors/functions/builtInFunctions.context.js]: Fixed

File: test/specs/parsing/functions/builtinFunctions.context.tests.js (hunk 5 above)
Comment (ThomasGravekamp) [inline test/specs/parsing/functions/builtinFunctions.context.tests.js]: Done

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+	return dayTimeDuration1._seconds < dayTimeDuration2._seconds;
+};
+
+/**
+ * @static
+ * @param   {DayTimeDuration}  dayTimeDuration1
+ * @param   {DayTimeDuration}  dayTimeDuration2
+ * @return  {boolean}
+ */
+DayTimeDuration.greaterThan = function (dayTimeDuration1, dayTimeDuration2) {
```
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: Done

File: src/selectors/dataTypes/valueTypes/DayTimeDuration.js [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+	constructor (seconds) {
+		// if (seconds > Number.MAX_SAFE_INTEGER || Math.abs(seconds) === Infinity) {
+		// 	throw new Error('FODT0002: Value overflow while constructing xs:yearMonthDuration');
+		// }
+		// this._seconds = seconds < Number.MIN_SAFE_INTEGER || Object.is(-0, seconds) ? 0 : seconds;
+		this._seconds = seconds;
+	}
+
+	getYears () {
+		return 0;
```
Comment (ThomasGravekamp) [inline src/selectors/dataTypes/valueTypes/DayTimeDuration.js]: Done

File: src/selectors/functions/builtInFunctions.numeric.js (hunk 1 above)
Comment (ThomasGravekamp) [inline src/selectors/functions/builtInFunctions.numeric.js]: Done

**Comment by ymor** [comment]
Hi,
Is this pull request still been worked on?  Would love to use this library but the lack of date/time functions is a bit of a blocker for us.   We are happy to help if required.
Thanks.

**Comment by DrRataplan** [comment]
Hy,

Work on this PR is paused because of changing priorities in our team. The main missing part is working with dates (addition and subtraction) and some wrong error codes which are thrown when intentionally non-implemented features are used. Casting works, most comparisons work, as well as most functions the spec defines.
This PR is also a bit out-of-date, we have done a couple of refactorings on the main branch which should be integrated.

I'll try to see if I can at least update this PR and make an inventory of the missing features.

Could you tell me which parts of the spec you'd like to use? After the rebase, I'd be happy to work with you in implementing the missing parts!

**Comment by ymor** [comment]
The main things we are looking for at the moment are the Comparison and Arithmetic operators on date, time and datetime.
Thanks again.

**Comment by DrRataplan** [comment]
It took some time, but I've rebased the feature branch. The tests fail because of some unrelated issue. Locally, they're OK.

The majority of the datetime related functions are implemented correctly, and the comparisons also seem to pass most of the QT3 tests (save for some obscure edge-cases). The main missing part is arithmetic on dates, times, etc. [The spec](https://www.w3.org/TR/xpath-functions-31/#dateTime-arithmetic) is a bit hard to read on this subject and some of the operators didn't map nicely on JavaScript dates, but I think I should be able to figure it out. More on this later this week.

BTW, thanks for bringing this PR back to my attention!

**Comment by DrRataplan** [comment]
I have finished refactoring the roughest edges of this PR, It is now in a state I'd like to integrate.

This PR brings the following:

- Support for the xs:date, xs:dateTime, xs:time, xs:yearMonthDuration, xs:dayTimeDuration
- The casting functions related to these types
- A number of functions related to these types
- Comparing these dates
- Arithmetic on durations
- Partial arithmetic on dates

What does this PR not bring:
 - Adding and subtracting dates and durations, these turned out to be more time consuming then I have available right now. These JavaScript functions throw a 'Not Implemented' error.
- 100% spec compliance for the implemented features: There are a number of failing QT3 tests, mostly around precision.

Next steps:
- I'll integrate this pull request, probably sometime today or next week.
- I do not have time to implement the missing features, but I will be happy to assist with implementing it and creating a pull request for it.

The operators which have not been implemented are located at https://github.com/FontoXML/fontoxpath/blob/implement-datetime-functions/src/selectors/dataTypes/valueTypes/DateTime.js#L444. The main difficulty I ran into is the overflow behaviour of JavaScript dates differs too much from how the spec defines it.
